### ISSUE TITLE
gmm weights changes not taken into account

### ISSUE THREAD

**Issue by DavidDoukhan** [idiap/bob#220]
Hi list,
I was willing to use class bob.learn.em.GMMMachine.
I was following the following documentation, which consist to create a GMMMachine instance, and set means, covariance and weights manually:
https://www.idiap.ch/software/bob/docs/latest/bioidiap/bob.learn.misc/master/guide.html

I installed bob with default options using pip.
According to the few logs bellow, it seems the GMM weights are not taken into account when computing the log likelihood of a sample.

```
import bob.learn.em
from bob.learn.em import GMMMachine
import numpy as np
print bob.learn.em.__version__

```
2.0.8
```
gmm1 = GMMMachine(1,1)
data = np.ones((1,1))
gmm1(data)
```
-1.4189385332046727
```
gmm1.weights[:] = 666
gmm1(data)
```
-1.4189385332046727

After few more investigation, I had the feeling that this behavior had to do with the caching of the log weights, which is not done when assigning new values to the weights, so I found a workaround consisting to create a new instance from a illed instance:

```
gmm2 = GMMMachine(gmm1)
gmm2(data)
```
5.082351137335717

This workaround seems to provide coherent results.

In conclusion, I think that this behavior should either be fixed, or set explicit within the documentation.

Regards,

**Comment by siebenkopf** [comment]
Hmm... I don't think that the weight assignment that you do is supported in the way you do it. Let me explain that a bit:

The GMM itself is implemented in C++. When you call the ``gmm1.weights``, internally the numpy array that wraps the weights in C++ is created and returned. Internally, this calls ``GMMMachine::getWeights``.
So, you are calling the getter function for the weights. Unfortunately, the update of the log weights is done in the **setter** function. So, you got the weights, you modified them, but the setter function is never called. 

A workaround is to actually call the setter function by assigning something to the GMM weights, instead of modifying the weights, i.e., ``gmm1.weights = numpy.array([666.])`` should do the trick.

Unfortunately, there is no way to actually do what you want to do. The only thing that we **might** implement is to make the getter function to return a const array, so that you will get an error message when you try to modify it directly.

**Comment by DavidDoukhan** [comment]
Thanks for your explanation! I was not aware of the setter behavior!

**Comment by siebenkopf** [comment]
In fact, we already return a const numpy array in the getter function:
https://github.com/bioidiap/bob.learn.em/blob/master/bob/learn/em/gmm_machine.cpp#L284

I am a bit surprised that you can actually modify the weights. 
@anjos: I have double-checked that the ``PyBlitzArrayCxx_AsConstNumpy`` function actually sets the ``writable`` flag to 0:
https://github.com/bioidiap/bob.blitz/blob/master/bob/blitz/include/bob.blitz/cppapi.h#L173
However, the ``PyBlitzArray_NUMPY_WRAP`` does not seem to copy that flag:
https://github.com/bioidiap/bob.blitz/blob/master/bob/blitz/api.cpp#L980
Hence, it seems to be possible to modify the array, although it should be const. Have you ever tested that the const numpy array wrapper that we return can not be changed?
